xilem/examples/memoization.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Arc;
+use xilem::view::{button, flex, memoize};
+use xilem::{AnyMasonryView, MasonryView, Xilem};
+
+// There are currently two ways to do memoization
+
+struct AppState {
+    count: i32,
+    // When TAITs are stabilized this can be a non-erased concrete type
+    count_view: Option<Arc<dyn AnyMasonryView<AppState>>>,
```

Comment:
Is it worth making a generic type which encapsulates this?
I.e. which owns the count and the `Option<Arc<V>>`

And so when the data is updated, it invalidates the view.

Because it seems incredibly easy to forget otherwise. IMO, that seems like a better fit for the Xilem threading model, although it's complicated

Reply:
Yeah agree. But I'm not sure if I understand you completely, do you mean something like in the last commit?

Reply:
So basically `Memoize` but with an `Arc` instead? I'm still not quite sure what you mean TBH.

So something like

```rust
struct MemoizedArcView<F, D, State> {
    data: D,
    // When TAITs are stabilized this can be a non-erased concrete type
    view: Option<Arc<dyn AnyMasonryView<State>>>,
    view_factory: F,
}

impl<State, D: PartialEq, F: Fn(&D) -> Arc<dyn AnyMasonryView<State>>>
    MemoizedArcView<F, D, State>
{
    fn view(&mut self) -> Arc<dyn AnyMasonryView<State>> {
        if let Some(view) = &self.view {
            view.clone()
        } else {
            let view = (self.view_factory)(&self.data);
            self.view = Some(view.clone());
            view
        }
    }

    fn set_data(&mut self, data: D) {
        self.data = data;
        self.view = None;
    }
}

fn memoized_arc_view<State, D, F>(data: D, view_factory: F) -> MemoizedArcView<F, D, State> {
    MemoizedArcView {
        data,
        view: None,
        view_factory,
    }
}
```
?

Though, I think having more direct control over invalidation with an `Arc` is an advantage over `Memoize` (not that this utility you're talking about is mutually exclusive though).

Reply:
This isn't quite what I was trying to describe, but that's probably better as a follow-up (probably related to #235)

Reply:
Yes might be, alright, keeping it in the example for now, we can still refactor later...